### ISSUE TITLE
pypa/build on Linux extension modules in purelib

### ISSUE THREAD

**Issue by tcuongd** [pypa/build#346]
Hello! I'm building a package on Linux with extension modules, using `setuptools` + `wheel`, and `pypa/build` to run the build command. I've added a custom distribution class to force platform-specific wheels:

```python
class BinaryDistribution(Distribution):
    """Distribution which always forces a binary package with platform name"""
    def has_ext_modules(self):
        return True

    def is_pure(self):
        return False

setup(..., distclass=BinaryDistribution)
```

This successfully generates a Linux-specific wheel. Now I want to "repair" the wheel in order to publish it to PyPi. I'm using `pypa/auditwheel` for this, but it gives this error:

```
RuntimeError: Invalid binary wheel, found the following shared library/libraries in purelib folder:
  	prophet_model.bin
  	diagnose
  	print
  	stanc
  	stansummary
  	diagnose.o
  	print.o
  	stansummary.o
  	bernoulli
  	main.o
The wheel has to be platlib compliant in order to be repaired by auditwheel.
```

I had initially thought it could be an issue with auditwheel (posted a question here: https://github.com/pypa/auditwheel/issues/333#issue-982011357) but I noticed that auditwheel is simple checking the metadata of the wheel here. I'm wondering if there's actually an issue in the wheel creation that's causing the extension modules to land in `purelib`. Would really appreciate any tips / where to start looking to solve this!

**Comment by henryiii** [comment]
> Linux with extension modules

If you have extension modules, setuptools will always make this platform specific. I'm guessing you are trying to add them manually instead of using setuptools? If that's the case, I'd generally override the extension build mechanism rather than distclass, such as in https://github.com/pybind/python_example/blob/master/setup.py

**Comment by FFY00** [comment]
This shouldn't be an issue with pypa/build as we simply provision the environment and invoke the backend, we take no part in the wheel building itself.

Could you provide more details on what you are doing, and perhaps a reproducible too? It seems to you are missing patching in someplace.

**Comment by FFY00** [comment]
I would transfer this to the setuptools repo, but I can't. @jaraco could you? Thanks :blush:

**Comment by henryiii** [comment]
Should this go to packaging-problems, perhaps?

**Comment by FFY00** [comment]
Well, I guess it is close enough, but this really is about customizing the setuptools build.

**Comment by henryiii** [comment]
I don't think this is a bug in setuptools, but rather a question on how to do something slightly tricky in packaging (and it touches several projects, auditwheel + setuptools at least).

**Comment by FFY00** [comment]
I guess, but the root issue really is setuptools, auditwheel is just complaining about a bad wheel. Feel free to move it to packaging-problems, it is not that clear where this belongs.

**Comment by henryiii** [comment]
Let's start there. I think we should first assume it's a user issue, and if it turns out to be a setuptools bug, it can move there. Same thing with issues about build, most of them start in packaging-problems and never need to move, as they are user issues with setup.py/setup.cfg. Once in a while, they are bugs in build, and then we move them here.
